File: 240202/아름다운 수/beautiful-number.py
n = int(input())    # n자리 수

selected = []
ans = 0
def Verify():       # 아름다운 수인지 판별
    curIdx, curNum = 0, selected[0]     # 첫번째 수부터 확인
    while True:
        if curIdx >= n:     # 만약 현재 보고 있는 인덱스가 범위를 벗어나면 break
            break

        curNum = selected[curIdx]   # 현재 확인하려는 수

        for i in range(curIdx+1, curIdx+curNum):    # 해당 숫자만큼 연달아 같은 숫자가 나오는지 확인
            if i >= n or curNum != selected[i]:     # 확인 도중 범위를 벗어나거나, 다른 숫자가 나온다면
                return False                        # 아름다운 수가 아님
        curIdx += curNum        # 다음 인덱스 이동시키기
        
    return True


def Choose(idx):    # idx번째 수를 선택하는 함수
    global ans
    if idx == n:
        if Verify():
            ans += 1
        return

    for i in range(1, 5):
        selected.append(i)
        Choose(idx+1)
        selected.pop()

Choose(0)
print(ans)

File: 240307/아름다운 수/beautiful-number.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> selected;
int n;
int ans;
void Check(){
    bool IsBeauty = true;
    
    int idx = 0, curNum = selected[idx];
    while(idx < (int)selected.size()){
        if(idx+curNum-1 >= n){
            IsBeauty = false;
            break;
        }

        for(int i=idx; i < idx+curNum; i++){
            if(curNum != selected[i])
                IsBeauty = false;
        }
        if(!IsBeauty)
            break;
        idx += curNum;
        curNum = selected[idx];
    }
    if(IsBeauty)
        ans++;

}

void Choose(int num){
    if(num == n){
        Check();
        return;
    }

    for(int i=1; i<=4; i++){
        selected.push_back(i);
        Choose(num+1);
        selected.pop_back();
    }
}

int main() {
    cin >> n;
    Choose(0);      // 0번째 숫자를 고르는 함수
    
    cout << ans;
    return 0;
}

/*
 *해설 풀이

#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int n;
int ans;
vector<int> seq;

bool IsBeautiful(){
    // 연달아 같은 숫자가 나오는 시작 위치를 잡는다
    for(int i=0; i<n; i += seq[i]){
        // 만약 연속하여 해당 숫자만큼 나올 수 없다면 아름다운 수가 아니다
        if(i + seq[i] - 1 >= n)
            return false;

        // 연속하여 해당 숫자만큼 같은 숫자가 있는지 확인한다.
        // 하나라도 다른 숫자가 있다면 아름다운 수가 아니다
        for(int j=i; j < i + seq[i]; j++)
            if(seq[j] != seq[i])
                return false;
    }
    return true;
}

void CountBeautifulSeq(int cnt){
    if(cnt == n){
        if(IsBeautiful())
            ans++;
        return;
    }

    for(int i=1; i<=4; i++){
        seq.push_back(i);
        CountBeautifulSeq(cnt+1);
        seq.pop_back();
    }
}

int main(){
    cin >> n;

    CountBeautifulSeq(0);

    cout << ans;
    return 0;
}
*/